File: 소수 찾기.py
# https://programmers.co.kr/learn/courses/30/lessons/42839

from itertools import permutations

def solution(numbers):
    answer = 0

    numbers = list(numbers)
    per_list = []

    for i in range(1,len(numbers)+1):
        for j in list(permutations(numbers,i)):
            a = int("".join(j))
            if a not in per_list:
                per_list.append(a)

    print(per_list)

    for num in per_list:
        check = True
        if num == 0 or num == 1:
            continue

        for c in range(2,int(num**0.5)+1):
            if num%c == 0:
                check = False
        
        if check:
            answer+=1

    return answer

numbers = "011"
print(solution(numbers))


from itertools import permutations

def solution1(n):
    a = set()
    for i in range(len(n)):
        a |= set(map(int, map("".join, permutations(list(n), i + 1))))  #문자열 조합으로 만들 수 있는 숫자를 set에 할당한다(중복 자동제거)
        print(a)

    a -= set(range(0, 2)) #0,1을 제거

    #가장 큰수에 대한 2 ~ a**0.5 사이 정수 범위를 만들고, 범위내의 각각의 정수값을 1step으로하는 배수set을 만들어, 그 배수set에 해당하는 값을 a에서 제거한다
    for i in range(2, int(max(a) ** 0.5) + 1):
        print(set(range(i * 2, max(a) + 1, i)))
        a -= set(range(i * 2, max(a) + 1, i))
        print(a)
    return len(a)


numbers = "011"
print(solution1(numbers))

File: 소수 찾기.js
// https://programmers.co.kr/learn/courses/30/lessons/12921

function solution(n) {
  var answer = 0;

  let prime = [2];
  let primeLen = 1;

  for (let i = 3; i <= n; i++) {
    let isPrime = true;
    let rootN = parseInt(i ** 0.5);
    // console.log(i,rootN);
    for (let j = 0; prime[j] <= rootN; j++) {
      //rootN보다 같거나 작은 소수들로 나눠서 확인해봄
      // console.log(i,prime[j]);
      if (i % prime[j] === 0) {
        isPrime = false;
        break;
      }
    }

    if (isPrime) {
      prime.push(i);
      primeLen += 1;
    }
    // console.log(prime)
  }
  return prime.length;
}

//다른답
function solution(n) {
  const s = new Set();
  //n까지의 홀수들을 가져옴
  for (let i = 1; i <= n; i += 2) {
    s.add(i);
  }
  s.delete(1);
  s.add(2);
  //3~sqrt(n)까지의 자연수를 배수로하는 s내의 값들을 순차적으로 지워나감 (채거르기 방식)
  for (let j = 3; j < Math.sqrt(n); j++) {
    //set.has(value) – 셋 내에 값이 존재하면 true, 아니면 false를 반환
    if (s.has(j)) {
      for (let k = j * 2; k <= n; k += j) {
        s.delete(k);
      }
    }
  }
  return s.size;
}
